File: Centroid Tracker.py
from collections import OrderedDict
import math

class CentroidTracker():

    def __init__(self , maxDisappeared = 5 ,  maxDistance = 80 ):
        self.nextObjectID = 0;
        self.objects = OrderedDict()
        self.disappeared = OrderedDict()
        self.maxDisappeared= maxDisappeared
        self.maxDistance  = maxDistance

        
    def register(self  , centroid ):
        self.objects[self.nextObjectID] = centroid
        self.disappeared[self.nextObjectID] = 0
        self.nextObjectID += 1
        
    def deregister(self , objectID ):
        del self.objects[objectID]
        del self.disappeared[objectID]

    def D( self , a , b , c , d ):
        return math.sqrt((a-c )*(a-c) + (b-d)*(b-d))

    def update(self , rects ) :
        if(len(rects) == 0):
            for objectID in list(self.disappeared.keys()):
                self.disappeared[objectID] += 1
                if self.disappeared[objectID] > self.maxDisappeared:
                    self.deregister(objectID)
            return self.objects

        inputCentroids = []
        for( i , (x , y , x1 , y1 )) in enumerate(rects):
             cx = ( x + x1)//2
             cy = (y + y1 )//2
             inputCentroids.append((cx , cy ))

        if( len(self.objects)==0):
             for i in  inputCentroids :
                self.register(i)
        else:
             objectCentroids = list(self.objects.values())
             idx = []
             dist = OrderedDict()
             for i , x in self.objects.items():
                 ans = (-1 , 99999999999 )
                 for j in range( 0 , len(inputCentroids)):
                     d =  self.D( x[0] , x[1], inputCentroids[j][0] , inputCentroids[j][1])
                     if( d <= ans[1]):
                           ans = ( j , d )
                 idx.append(( i , ans[0]))
                 dist[i] = int(ans[1])

             used_idx = set()
             used_current_cent= set()

             unused = set( range(0 , len(inputCentroids)))
             unused_cent = set(self.objects.keys())

             for i in range( 0 , len(idx)):
                if idx[i][0] in used_idx or idx[i][1] in used_current_cent :
                    continue

                if dist[idx[i][0]] > self.maxDistance :
                    continue

                self.objects[idx[i][0]] = inputCentroids[idx[i][1]]
                self.disappeared[idx[i][0]] = 0

                used_idx.add(idx[i][0])
                used_current_cent.add(idx[i][1])

                if idx[i][0] in unused_cent :
                    unused_cent.remove(idx[i][0])
                if idx[i][1] in unused :
                    unused.remove(idx[i][1])


           
             if(len(objectCentroids) >= len(inputCentroids)):
                 for i in unused_cent:
                     self.disappeared[i]+= 1
                     if self.disappeared[i] > self.maxDisappeared :
                         self.deregister(i)
             else:
                 for i in unused :
                     self.register(inputCentroids[i])                     
        return self.objects

                         
                     
             
             



import cv2
face_cascade = cv2.CascadeClassifier('p1.xml') 
cap = cv2.VideoCapture(0) 
ct = CentroidTracker() 
x = 0
while 1:   
    ret, img = cap.read()  
   
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY) 
    faces = face_cascade.detectMultiScale(gray, 1.3, 5) 
    rect = []
    for (x,y,w,h) in faces: 
        rect.append( (x ,y , x +w , y + h) )
       
    objects = ct.update(rect)
    for (x,y,w,h) in faces: 
        cv2.rectangle(img,(x,y),(x+w,y+h),(255,255,0),2)
        
    for (objectID, centroid) in objects.items():
        text = "ID {}".format(objectID)
        cv2.putText(img, text, (centroid[0] - 10, centroid[1] - 10),cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)
        cv2.circle(img, (centroid[0], centroid[1]), 4, (0, 255, 0), -1)

    cv2.imshow('img',img)
    k = cv2.waitKey(1) & 0xff
    if k == 27: 
        break
cap.release() 

File: Centroid Tracker.cpp
#include"iostream"
#include "map"
#include "set"
#include "vector"
#include "math.h"

using namespace std;

struct Rect{
    int x ;
    int y ;
    int width ;
    int height;
};




class Trackableobject{
   public:
  	int obj_ide;
  	vector<pair<int , int >> centroid ;
  	int count ;
  public :
  	Trackableobject( int objectID , pair<int , int > centroid );
 };


Trackableobject::Trackableobject( int objectID , pair<int , int > cent_point ){
           obj_ide = objectID;
           centroid.push_back(cent_point);
           count = 0 ;
  }



class My_Tracker{

private:
	int next_ide = 0;
	map<int , pair<int , int > > objects;
	std::map<int , Trackableobject* > mp;
	map<int , int >disappeared;
        vector<pair<int , pair<int , int >>>idx;
 	map<int , int  >dist;
        vector<pair<int , int >>inputCentroid ;
 	set<int >unused_current_centroid;
 	set<int >unused_centroid;
        set<int>used_idx;
        set<int>used_current_centroid;


	int mxDisappeared  = 50;
	int mxDistance = 40;
        int W = 500;
        int H = 150;
        pair<int , int >check_line_cross;
        int total_up =0;
  	int total_down =0;
public :
    My_Tracker(int maxDisappeared  ,  int maxDistance  );
    void registered(pair<int , int > centroid );
    void deregistered( int objectID);
    int D( int a , int b , int c , int  d);
    void update(std::vector<cv::Rect>v);
    void draw(cv::Mat &img );
    void check();
};

My_Tracker::My_Tracker(int maxDisappeared ,  int maxDistance  ){
	//My_Tracker::mxDisappeared = mxDisappeared;
	//My_Tracker::mxDistance = mxDistance ;
}

 void My_Tracker::registered(pair<int , int > centroid ){
 	objects[next_ide] = centroid;
 	disappeared[next_ide] = 0;
 	++ next_ide;
 }

void My_Tracker::deregistered( int objectID){
        auto it = objects.find(objectID);
        if(it != objects.end()) objects.erase(it);
        auto it_1 = disappeared.find(objectID);
	if(it_1 != disappeared.end()) disappeared.erase(it_1);
	if(mp.find(objectID) != mp.end()) mp.erase(objectID);
}


 int   My_Tracker:: D( int a , int b , int c , int  d){
 	return sqrt((a-c)*(a-c) + (b-d)*(b-d));
 }

 void  My_Tracker:: update(std::vector<Rect>v){

 	if(v.size()==0){
 		for(auto it : disappeared){
 			disappeared[it.first] += 1;
 			if(disappeared[it.first] >mxDisappeared)
 				 deregistered(it.first);
 		}
 		return ;
 	}

 	inputCentroid.clear() ;
 	unused_current_centroid.clear();
 	unused_centroid.clear();
        idx.clear();
 	dist.clear();

 	for(int i = 0 ; i<v.size() ; ++i ){
        Rect tmp = v[i];
        int cx = (tmp.x +  (tmp.x +  tmp.width ))/2;
        int cy = (tmp.y +  (tmp.y + tmp.height ))/2;
        inputCentroid.push_back({cx , cy });
        unused_current_centroid.insert(i);
 	}

 	if(objects.size()==0){
              for(auto it : inputCentroid)
                   registered(it);
 	}else{

 	   
 	   for(auto it : objects ){
               pair<int , int >ans = {-1 , INT_MAX};
               for( int j = 0 ; j<inputCentroid.size() ; ++j ){
               int  d = D(it.second.first , it.second.second , inputCentroid[j].first , inputCentroid[j].second );
               if( d < ans.second ){
                ans = { j , d };
             }
          }
           idx.push_back({ans.second ,{it.first , ans.first }});
           
           if(dist[it.first] > ans.second && dist[it.first] ){
            dist[it.first] = ans.second ;
           
           
           }
           if(dist[it.first] == 0 ){
                 dist[it.first] = ans.second ;
           }
           unused_centroid.insert(it.first);
 	}
        
        sort(idx.begin() , idx.end());
 	used_idx.clear();
 	used_current_centroid.clear();

       for(auto it : idx ){

           if(used_idx.find(it.second.second)!= used_idx.end()|| used_current_centroid.find(it.second.first)!= used_current_centroid.end())
                    continue;

           if(dist[it.second.first] > mxDistance ){
                   continue;
           }

         objects[it.second.first] = inputCentroid[it.second.second ];
         disappeared[it.second.first] = 0;

        used_idx.insert(it.second.second);
        used_current_centroid.insert(it.second.first);

        auto x = unused_current_centroid.find(it.second.second);
        if( x  != unused_current_centroid.end())unused_current_centroid.erase(x);

         x = unused_centroid.find(it.second.first);
        if(x != unused_centroid.end()) unused_centroid.erase(x);
       }

       if(objects.size() >= inputCentroid.size()){
        for(auto it : unused_centroid ){
            disappeared[it] += 1;
            if( disappeared[it] > mxDisappeared )
                deregistered(it);
        }
       }else{
         for(auto it : unused_current_centroid)registered(inputCentroid[it]);
       }
 	}
    return ;
 }





  void My_Tracker :: draw(cv::Mat &img ){
	    for(auto it : objects ){
	   	string s = "ID " + to_string(it.first);
	   	cv::putText( img , s , cv::Point(it.second.first -10 , it.second.second-10) , FONT_HERSHEY_DUPLEX, 0.5 ,cv::Scalar(0, 255, 0) , 2);
	   	cv::circle(img , cv::Point(it.second.first , it.second.second ) , 4 ,cv::Scalar(0 , 255 , 0 ) , -1 );
	  }

	  string s1 = "OUT : " + to_string(total_up) ;
          string s2 = " IN : " + to_string(total_down);
	  cv::putText( img , s1  , cv::Point(50 , H-20) , FONT_HERSHEY_DUPLEX, 1 ,cv::Scalar(0, 0, 255) , 2);
          cv::putText( img , s2  , cv::Point(50 , H+30) , FONT_HERSHEY_DUPLEX, 1 ,cv::Scalar(0, 0, 255) , 2);
	  if(check_line_cross.first==1 || check_line_cross.second ==1){
	          cv::line(img , cv::Point(0 , H) , cv::Point( W , H ) , Scalar(0, 0 , 255) , 2 );
          }
	  else{
	  	  cv::line(img , cv::Point(0 , H ) , cv::Point(W , H ) , Scalar(0, 255, 255) , 2 );
              }

  }

int main(){


}
